Added integration test using kitchen and vagrant.

jperville opened the pull request:
Hello @aespinosa,

This is a first PR to implement real integration tests for this cookbook using kitchen, vagrant and inspec.
I fixed a few details (such as the docker repository in the test suite) to make it all work.

In the next days, I would like to submit PRs to add support for Ubuntu 16.04 and to support kubernetes 1.2.2 (through refactoring of URL/checksum of artefacts into attributes). The new PRs will require this PR to be merged first.

Thank you very much for this great cookbook!

jperville commented:
Please note that the "bundle exec rake test" task does not work for me because of the "cheffish" gem which requires "chef", I am having circular dependencies issues. Could you complete the Gemfile with the versions you are using?

aespinosa commented:
Thanks for adding the tests. I don't have a Gemfile as I use nix to isolate my dependencies.

I don't think my unit tests need berkshelf. It's probably better to have test-kitchen and family to be in a separate gem_home/ omnibus environment and shell it out.

aespinosa commented on test/integration/default/inspec/assert_functioning_spec.rb:
```diff
@@ -0,0 +1,83 @@
+############################
+
+# etcd
+describe service('etcd-default') do
```
we probably don't need to test if etcd and flannel are working.

jperville commented:
Thanks for reviewin my PR! I will submit an updated version in a few days with all your comments.

aespinosa commented on .kitchen.yml:
```diff
@@ -1,6 +1,15 @@
 ---
+driver:
```
if you put this in `~/.kitchen/config.yml` instead you won't need this in the rest of your kitchen file.  

I'm actually working on setting up CI servers to spin up instances on the cloud. I prefer not having this here so I don't have to populate my CI jobs with `.kitchen.local.yml` files

aespinosa commented on Rakefile:
```diff
@@ -1,16 +1,42 @@
+# Integration tests. Kitchen.ci
+namespace :integration do
+  desc 'Run Test Kitchen with Vagrant'
+  task :vagrant do
```
How different is this from running the kitchen:all task?

aespinosa commented:
Output with `verifier.format: doc`:

I tend to be picky with rspec output making sure it reads like real sentences.  I'll comment on some of the context inline.

But this is a good start with everything passing! :)

```
Service etcd-default
  should be installed
  should be enabled
  should be running

Port  4001
Could not parse fe80::42:8dff:fe4b::123, bad URI(is not URI?): addr://[fe80::42:8dff:fe4b:]:123
  should be listening
  processes
    should include "etcd"

Service flannel-default
  should be installed
  should be enabled
  should be running

File /run/flannel/subnet.env
  should exist

Service kube-apiserver
  should be installed
  should be running

Port  8080
Could not parse fe80::42:8dff:fe4b::123, bad URI(is not URI?): addr://[fe80::42:8dff:fe4b:]:123
  should be listening
  processes
    should include "kube-apiserver"

Port  6443
  should be listening
  processes
    should include "kube-apiserver"

Service kube-controller-manager
  should be installed
  should be running

Service kube-proxy
  should be installed
  should be running

Service kube-scheduler
  should be installed
  should be running

Service kubelet
  should be installed
  should be running

Port  10250
Could not parse fe80::42:8dff:fe4b::123, bad URI(is not URI?): addr://[fe80::42:8dff:fe4b:]:123
  should be listening
  processes
    should include "kubelet"

Command kubectl get node
  exit_status
    should eq 0
  stdout
    should match /default.*Ready/

Command kubectl get pod busybox
  exit_status
    should eq 0
  stdout
    should match /busybox.*Running/

Finished in 0.19414 seconds (files took 0.39012 seconds to load)
29 examples, 0 failures
```

aespinosa commented on test/integration/default/inspec/assert_functioning_spec.rb:
```diff
@@ -0,0 +1,83 @@
+  it { should exist }
+end
+
+# kube-apiserver
```
Instead of comments, let's move this into its own context

jperville commented:
Thanks for the code review @aespinosa, I will apply all your remarks later this weekend.

Would you rather have separate commits added to this PR that you can merge after squashing or should I git push --force to keep this PR in a single commit?

aespinosa commented:
I prefer them to be in a single commit. thanks @jperville !

jperville commented:
I will be using forced push on this branch to have all changes in a single commit then, thanks @aespinosa .

jperville commented:
I pushed a new version of this PR will the following changes:
 - removed etcd and flannel tests
 - moved 'berkshelf' dependency to the kitchen-common group (not needed for unit tests)
 - commented the vagrant driver in ~/kitchen.yml
 - use Kitchen::Raketasks in Rakefile to export integration tests
 - integration test now outputs in 'doc' format by default

I was not able to use contexts to group the assertions by component because the syntax is not supported by inspec yet (see https://github.com/chef/inspec/issues/630). The use of 'controls' could work but it will not display anything in the 'doc' output.

Please let me know if there are other improvements you would like to see in this PR.

aespinosa commented:
I'll try messing with the spec descriptions when I have time. 

This is a good start to make sure upgrade paths are working. Thanks for the contribution!

jperville commented:
Thank you very much for merging @aespinosa.
